kernel/src/hil/asymmetric_encryption.rs

```diff
@@ -0,0 +1,133 @@
+//! Interface for Public/Private key encryption
+
+use crate::hil::entropy;
+use crate::ErrorCode;
+
+/// Upcall from the `PubPrivKey` trait.
+pub trait PubPrivKeyClient<'a> {
+    /// The `generate()` command has been completed.
+    fn generation_complete(&'a self, result: Result<(), ErrorCode>);
+}
+
+/// An internal representation of a asymetric Public and/or Private key.
+///
+/// This trait is useful for managing keys internally in Tock.
+pub trait PubPrivKey<'a> {
+    /// Set the client. This client will be called when the `generate()`
+    /// function is complete. If using an existing key this doesn't need to be
+    /// used.
+    fn set_client(&'a self, client: &'a dyn PubPrivKeyClient<'a>);
+
+    /// This generates a new private/public key pair of `length` bits.
+    /// This will call the `generation_complete()` on completion. They keys
+    /// cannot be used and will return `None` until the upcall has been called.
+    ///
+    /// The keys generated by `generate()` will depend on the implementation.
+    fn generate(&'a self, trng: &'a dyn entropy::Entropy32, length: usize)
+        -> Result<(), ErrorCode>;
+
+    /// Import an existing public key. This will override any existing keys.
+    fn import_public_key(&mut self, public_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Import an existing private key. This will override any existing keys.
+    fn import_private_key(&mut self, private_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Store the public key in buffer if it exists.
+    fn pub_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+
+    /// Store the private key in buffer if it exists.
+    fn priv_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+}
+
+pub enum Operation {
+    Encrypt,
+    Decrypt,
+    Sign,
+    Verify,
+}
+
+/// Upcall from the `RsaCrypto` trait.
+pub trait Client<'a, B> {
+    /// The `encrypt()`/`decrypt()`/`sign()` command has been completed.
+    ///
+    /// On success the value of `result` should be ignored unless the `op`
+    /// was a `Operation::Verify` in which cast this indicates if the message
+    /// is valid.
+    fn operation_done(&'a self, result: Result<bool, ErrorCode>, op: Operation, buffers: B);
```

Comment:
I'd suggest splitting encrypt/decrypt/sign/verify apart. With `dyn` references, you are going to get everything included even if you only use a small part of it. It might make sense to have encrypt/decrypt in one trait and sign/verify in another.

You want separate callbacks for the different operations. Having a shared structure for buffers, whose fields have different meanings based on the operation, is not taking advantage of type safety. This looks very C-like. 

Why not have an RsaClient? The major advantage of this approach would be that you could write generic modules using client that are parameterized by B. But since this generic isn't in the opposite direction (you have an RSA trait), I'm unsure of when this would be useful.

Reply:
I have changed `Client` to be `RsaClient`.

I don't agree though that splitting the complete upcalls apart is better. If you look at the [capsule upcall code](https://github.com/tock/tock/pull/2693/files#diff-6908c29af6e08ab6f8f3882b89317a78fcff3193b01ef0a033aa98bae1074956R200) we can cleanly pass the complete command up to userspace. Splitting these in the kernel would just make that clunky.

We could debate splitting the kernel and userspace upcalls, but I'm not convinced that gives us any benefits. Consolidating all of the callbacks into a single one to userspace I think is simpler.

Reply:
I think it's important to distinguish a kernel HIL and a userspace API. There are going to be users of this HIL within the kernel that are not a system call capsule. We want to design the HIL for internal kernel use, where *one* of the uses is a system call capsule. A HIL is intended to be general, not for a specific use case.

Reply:
I agree. I don't see why splitting it up will be easier be any better for internal kernel use either. The current way will work internally and works very well with the only current user which is the system call capsule.